flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReaderImplTest.java

```diff
@@ -0,0 +1,205 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.flink.core.memory.HeapMemorySegment;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.runtime.checkpoint.StateObjectCollection;
+import org.apache.flink.runtime.checkpoint.TaskStateSnapshot;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.state.InputChannelStateHandle;
+import org.apache.flink.runtime.state.memory.ByteStreamStateHandle;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.HAS_MORE_DATA;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.NO_MORE_DATA;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * {@link ChannelStateReaderImpl} test.
+ */
+public class ChannelStateReaderImplTest {
+
+	private static final InputChannelInfo CHANNEL = new InputChannelInfo(1, 2);
+	private static final byte[] DATA = generateData(10);
+	private ChannelStateReaderImpl reader;
+
+	@Before
+	public void init() {
+		reader = getReader(CHANNEL, DATA);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		reader.close();
+	}
+
+	@Test
+	public void testDifferentBufferSizes() throws Exception {
+		for (int bufferSize = 1; bufferSize < 2 * DATA.length; bufferSize++) {
+			try (ChannelStateReaderImpl reader = getReader(CHANNEL, DATA)) { // re-create reader to re-read the same channel
+				readAndVerify(bufferSize, CHANNEL, DATA, reader);
+			}
+		}
+	}
+
+	@Test
+	public void testWithOffsets() throws IOException {
+		Map<InputChannelStateHandle, byte[]> handlesAndBytes = generateHandlesWithBytes(10, 20);
+		ChannelStateReader reader = new ChannelStateReaderImpl(taskStateSnapshot(handlesAndBytes.keySet()), new ChannelStateSerializerImpl());
+		for (Map.Entry<InputChannelStateHandle, byte[]> e : handlesAndBytes.entrySet()) {
+			readAndVerify(42, e.getKey().getInfo(), e.getValue(), reader);
+		}
+	}
+
+	@Test(expected = Exception.class)
+	public void testReadOnlyOnce() throws IOException {
```

Comment:
If possible, it is better to further verify the internal `RefCountingFSDataInputStream` is dereferenced and closed when no more data.  Then we can confirm no resource leak.